src/tests/dbus-tests/test_70_encrypted.py

```diff
@@ -609,6 +609,53 @@ def test_convert_xfail(self):
 
         self.assertEqual(2, _get_luks_version(disk))
 
+    def test_header_backup_restore(self):
+        disk = self.vdevs[0]
+        device = self.get_device(disk)
+        self._create_luks(device, self.PASSPHRASE)
+
+        self.addCleanup(self._remove_luks, device)
+        self.udev_settle()
+
+        # check that backup normally works
+        BACKUP_FILE = "/var/tmp/udisks_encrypted_header_backup.luks"
+        if os.path.exists(BACKUP_FILE):
+            os.unlink(BACKUP_FILE)
+        self.assertFalse(os.path.exists(BACKUP_FILE))
+
+        device.HeaderBackup(BACKUP_FILE, self.no_options,
+                            dbus_interface=self.iface_prefix + '.Encrypted')
+        self.assertTrue(os.path.exists(BACKUP_FILE))
+
+        # check that backup will fail if there is already a file under the specified path
+        msg = 'org.freedesktop.UDisks2.Error.Failed: Error backing up header of encrypted device /dev/.+: Failed to backup LUKS header: Invalid argument'
+        with self.assertRaisesRegex(dbus.exceptions.DBusException, msg):
+            device.HeaderBackup(BACKUP_FILE, self.no_options,
+                                dbus_interface=self.iface_prefix + '.Encrypted')
+
+        self.assertTrue(os.path.exists(BACKUP_FILE))
+
+        # check that after reaping device and restoring header, cryptsetup will recognize header
+        device.Lock(self.no_options, dbus_interface=self.iface_prefix + '.Encrypted')
+        ret, out = udiskstestcase.UdisksTestCase.run_command("cryptsetup luksDump %s" % disk)
+        self.assertEqual(ret, 0)
+        DISK_SIZE = 100 # magic number from `targetcli_config.json`
+        ret, out = udiskstestcase.UdisksTestCase.run_command("dd if=/dev/zero of=%s bs=1M count=%d" % (disk, DISK_SIZE))
+        self.assertEqual(ret, 0)
+        ret, out = udiskstestcase.UdisksTestCase.run_command("cryptsetup luksDump %s" % disk)
+        self.assertEqual(1, ret)
+        self.assertTrue(("Device %s is not a valid LUKS device." % disk) in out)
```

Comment:
There is a race condition here -- `dd` removes the header, but the change is not yet propagated from udev to udisks so the restore works even with the check for LUKS metadata in the code. You need to wait for the `IdType` property to become empty after calling `dd` with somethig like

```
fstype = self.get_property(device, '.Block', 'IdType')
fstype.assertEqual('')
```